File: TouchNothing/Views/LeaderboardScreen.swift
import SwiftUI

struct LeaderboardScreen: View {
    let onBack: () -> Void

    @State private var mode: LeaderboardMode = .bestSession
    @State private var leaderboardRows: [LeaderboardRow] = []
    @State private var statusMessage: String?
    @State private var isLoading = false
    @State private var loadFailed = false

    var body: some View {
        VStack(spacing: 0) {
            headerView
            modePicker
            statusBanner
            contentArea
            ScreenBackButton(action: onBack)
        }
        .background(AppColors.background)
        .statusBarHidden(true)
        .persistentSystemOverlays(.hidden)
        .forceHiddenStatusBar()
        .task(id: mode) {
            await loadLeaderboard()
        }
    }

    private var headerView: some View {
        Text(LocalizationKey.leaderboardTitle.localized)
            .font(.title2)
            .foregroundStyle(AppColors.textPrimary)
            .padding(.top, 24)
            .padding(.bottom, 12)
    }

    private var modePicker: some View {
        Picker("", selection: $mode) {
            ForEach(LeaderboardMode.allCases, id: \.self) { leaderboardMode in
                Text(leaderboardMode.titleKey.localized).tag(leaderboardMode)
            }
        }
        .touchNothingSegmentedPickerStyle()
        .padding(.horizontal, 16)
        .padding(.bottom, 12)
    }

    @ViewBuilder
    private var statusBanner: some View {
        if let statusMessage {
            Text(statusMessage)
                .font(.footnote)
                .foregroundStyle(AppColors.textSecondary)
                .padding(.horizontal, 16)
                .padding(.bottom, 8)
        }
    }

    @ViewBuilder
    private var contentArea: some View {
        if isLoading, leaderboardRows.isEmpty {
            LoadingStateView()
        } else if loadFailed {
            failureView
        } else if leaderboardRows.isEmpty {
            emptyView
        } else {
            leaderboardList
        }
    }

    private var emptyView: some View {
        VStack {
            Spacer()
            Text(LocalizationKey.leaderboardEmpty.localized)
                .foregroundStyle(AppColors.textSecondary)
            Spacer()
        }
    }

    private var failureView: some View {
        VStack {
            Spacer()
            Text(LocalizationKey.leaderboardError.localized)
                .foregroundStyle(AppColors.textSecondary)
                .multilineTextAlignment(.center)
                .padding(.horizontal, 16)
            Spacer()
        }
    }

    private var leaderboardList: some View {
        ScrollView {
            LazyVStack(spacing: 0) {
                ForEach(leaderboardRows) { row in
                    LeaderboardRowView(row: row)
                        .padding(.horizontal, 16)
                        .padding(.vertical, 12)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .background(rowBackground(for: row))
                }
            }
        }
        .refreshable {
            await loadLeaderboard()
        }
    }

    private func rowBackground(for row: LeaderboardRow) -> Color {
        row.isCurrentUser ? AppColors.fieldBackground : AppColors.background
    }

    private func loadLeaderboard() async {
        isLoading = true
        loadFailed = false
        defer { isLoading = false }

        let result = await LeaderboardService.load(
            mode: mode,
            currentNickname: LocalUserStore.nickname
        )
        leaderboardRows = result.rows
        statusMessage = result.statusMessage
        loadFailed = result.rows.isEmpty && result.statusMessage == LocalizationKey.leaderboardError.localized
    }
}

File: app/src/main/java/com/touchnothing/app/ui/screen/LeaderboardScreen.kt
package com.touchnothing.app.ui.screen

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Text
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.touchnothing.app.R
import com.touchnothing.app.data.model.LeaderboardMode
import com.touchnothing.app.data.model.LeaderboardRow
import com.touchnothing.app.domain.LeaderboardLoadResult
import com.touchnothing.app.domain.LeaderboardService
import com.touchnothing.app.ui.component.LeaderboardRowView
import com.touchnothing.app.ui.component.LoadingStateView
import com.touchnothing.app.ui.component.ScreenBackButton
import com.touchnothing.app.ui.component.TouchNothingScreenLayout
import com.touchnothing.app.ui.component.TouchNothingSegmentedPicker
import com.touchnothing.app.ui.theme.AppColors
import kotlinx.coroutines.launch

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun LeaderboardScreen(
    leaderboardService: LeaderboardService,
    currentNickname: String?,
    onBack: () -> Unit,
) {
    val context = LocalContext.current
    val refreshScope = rememberCoroutineScope()
    var mode by remember { mutableStateOf(LeaderboardMode.BEST_SESSION) }
    var rows by remember { mutableStateOf<List<LeaderboardRow>>(emptyList()) }
    var statusMessage by remember { mutableStateOf<String?>(null) }
    var isLoading by remember { mutableStateOf(false) }
    var isRefreshing by remember { mutableStateOf(false) }
    var loadFailed by remember { mutableStateOf(false) }
    var loadGeneration by remember { mutableIntStateOf(0) }

    val modeOptions = listOf(
        stringResource(R.string.leaderboard_mode_best_session),
        stringResource(R.string.leaderboard_mode_total),
    )
    val selectedModeIndex = if (mode == LeaderboardMode.BEST_SESSION) 0 else 1

    suspend fun applyLoadResult(result: LeaderboardLoadResult) {
        rows = result.rows
        statusMessage = result.statusMessage(context)
        loadFailed = result.isFailure
    }

    suspend fun loadLeaderboard(generation: Int) {
        val result = leaderboardService.load(mode, currentNickname)
        if (generation != loadGeneration) return
        applyLoadResult(result)
    }

    LaunchedEffect(mode) {
        loadGeneration += 1
        val generation = loadGeneration
        isLoading = true
        loadFailed = false
        loadLeaderboard(generation)
        isLoading = false
    }

    fun refreshLeaderboard() {
        refreshScope.launch {
            isRefreshing = true
            loadGeneration += 1
            val generation = loadGeneration
            loadLeaderboard(generation)
            isRefreshing = false
        }
    }

    Column(modifier = Modifier.fillMaxSize().background(AppColors.Background)) {
        TouchNothingScreenLayout {
            Column(modifier = Modifier.fillMaxWidth()) {
                Text(
                    text = stringResource(R.string.leaderboard_title),
                    fontSize = 22.sp,
                    color = AppColors.TextPrimary,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(top = 24.dp, bottom = 12.dp),
                    textAlign = TextAlign.Center,
                )
                TouchNothingSegmentedPicker(
                    options = modeOptions,
                    selectedIndex = selectedModeIndex,
                    onSelected = { index ->
                        mode = if (index == 0) LeaderboardMode.BEST_SESSION else LeaderboardMode.TOTAL_TIME
                    },
                )
                statusMessage?.let { message ->
                    Text(
                        text = message,
                        color = AppColors.TextSecondary,
                        fontSize = 12.sp,
                        modifier = Modifier.padding(top = 8.dp),
                    )
                }
            }
        }

        PullToRefreshBox(
            isRefreshing = isRefreshing,
            onRefresh = ::refreshLeaderboard,
            modifier = Modifier
                .weight(1f)
                .fillMaxWidth(),
        ) {
            Box(modifier = Modifier.fillMaxSize()) {
                when {
                    isLoading && rows.isEmpty() -> LoadingStateView(Modifier.align(Alignment.Center))
                    loadFailed -> Text(
                        text = stringResource(R.string.leaderboard_error),
                        color = AppColors.TextSecondary,
                        modifier = Modifier.align(Alignment.Center).padding(horizontal = 16.dp),
                        textAlign = TextAlign.Center,
                    )
                    rows.isEmpty() -> Text(
                        text = stringResource(R.string.leaderboard_empty),
                        color = AppColors.TextSecondary,
                        modifier = Modifier.align(Alignment.Center),
                    )
                    else -> LazyColumn(modifier = Modifier.fillMaxSize()) {
                        items(rows, key = { it.id }) { row ->
                            LeaderboardRowView(
                                row = row,
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .background(
                                        if (row.isCurrentUser) AppColors.FieldBackground else AppColors.Background,
                                    )
                                    .padding(horizontal = 16.dp, vertical = 12.dp),
                            )
                        }
                    }
                }
            }
        }

        ScreenBackButton(onClick = onBack)
    }
}
